extensions/omni_sql/omni_sql.c

```diff
@@ -606,147 +716,44 @@ Datum execute_parameterized(PG_FUNCTION_ARGS) {
-            switch (id) {
-            case INT2OID:
-              call_ctx->values[i] =
-                  Int16GetDatum(numeric_int4_opt_error(DatumGetNumeric(call_ctx->values[i]), NULL));
-              break;
-            case INT4OID:
-              call_ctx->values[i] =
-                  Int32GetDatum(numeric_int4_opt_error(DatumGetNumeric(call_ctx->values[i]), NULL));
-              break;
-            case INT8OID:
-#if PG_MAJORVERSION_NUM >= 17
-              call_ctx->values[i] =
-                  Int64GetDatum(numeric_int8_opt_error(DatumGetNumeric(call_ctx->values[i]), NULL));
-#else
-              call_ctx->values[i] = DirectFunctionCall1(numeric_int8, call_ctx->values[i]);
-#endif
-              break;
-            case FLOAT4OID:
-              call_ctx->values[i] = DirectFunctionCall1(numeric_float4, call_ctx->values[i]);
-              break;
-            case FLOAT8OID:
-              call_ctx->values[i] = DirectFunctionCall1(numeric_float8, call_ctx->values[i]);
-              break;
-            default: {
-              HeapTuple typeTuple = SearchSysCache1(TYPEOID, ObjectIdGetDatum(id));
-              if (!HeapTupleIsValid(typeTuple)) {
-                ereport(ERROR, (errmsg("cache lookup failed for type %u", id)));
-              }
-
-              char *typename = pstrdup(NameStr(((Form_pg_type)GETSTRUCT(typeTuple))->typname));
-
-              ReleaseSysCache(typeTuple);
-
-              ereport(ERROR, errmsg("can't convert numeric to type %s", typename));
-              break;
-            }
-            }
-          }
-          call_ctx->types[i] = id;
-          i++;
-        }
-        array_free_iterator(iter);
+        extract_information_json_types(call_ctx, PG_GETARG_ARRAYTYPE_P(2));
       }
     }
-
+    SPI_connect();
+    // Prepare the cursor
+    SPIPlanPtr plan = SPI_prepare(call_ctx->stmt, call_ctx->nargs, call_ctx->types);
+    // is this a plan that could return rows?
+    call_ctx->returning = SPI_is_cursor_plan(plan);
     if (call_ctx->returning) {
-      SPI_connect();
-
-      // Prepare the cursor
-      SPIPlanPtr plan = SPI_prepare(call_ctx->stmt, call_ctx->nargs, call_ctx->types);
       if (plan == NULL) {
         ereport(ERROR, errmsg("%s", SPI_result_code_string(SPI_result)));
       }
       call_ctx->portal =
           SPI_cursor_open("_omni_sql_execute", plan, call_ctx->values, call_ctx->nulls, false);
       // Fetch zero records just to get the tupdesc
       SPI_cursor_fetch(call_ctx->portal, true, 0);
-
       // Switch to multi-call memory instead of SPI
       MemoryContext oldcontext1 = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
       // Copy tupdesc
       funcctx->tuple_desc = CreateTupleDescCopyConstr(SPI_tuptable->tupdesc);
       BlessTupleDesc(funcctx->tuple_desc);
       funcctx->attinmeta = TupleDescGetAttInMetadata(funcctx->tuple_desc);
       MemoryContextSwitchTo(oldcontext1);
-
       SPI_finish();
     } else {
+      SPI_finish();
```

Comment:
Good catch. Technically speaking, we don't actually need both of the `SPI_finish()` below if we get a bit more clever below (continuing using the first connection on the first run), but it's not critical. 

What is critical is we must ensure that for every SPI_connect there will always be a matching SPI_finish in runtime. Which this bit helps with.

👍